### PR TITLE
[PREVIEW] MRU query strings and manifest-less queries

### PR DESCRIPTION
**Pull request by jpeletier**
Building on https://github.com/ethersphere/go-ethereum/pull/813, this PR makes the following changes:
Changed from path components to query strings:
`GET /bzz-resource:/<MANIFEST_ADDRESS>/?period=xx&version=xx` for specific period, version.
`GET /bzz-resource:/<MANIFEST_ADDRESS>/?period=xx` for specific period, latest version
`GET /bzz-resource:/<MANIFEST_ADDRESS>/` retrieves the latest, as before.

Added the following parameters allows to skip the manifest altogether by providing the info the manifest would provide:
`GET /bzz-resource:/?starttime=xx&frequency=xx&topic=0x000000...&user=0x00000...` 

you can also add `&period=xx&version=xx` to the above as well, to query a specific period / version.

Changed the way to create / update. You can now either:

POST `/bzz-resource:/<MANIFEST_ADDRESS>/?period=xx&version=xx&signature=0x0000...` (update data must be in the body of the request)

or skip the manifest by providing the info the manifest would provide:
`POST/bzz-resource:/?starttime=xx&frequency=xx&topic=0x000000...&owner=0x00000&period=xx&version=xx&signature=0x...`
again, update data must be in the body of the request.

If you want your node to create and publish a manifest along with the update, add the query flag `&manifest=1` to the above POST query.

Therefore, with this change POST/GET are symmetric and publishing a manifest is now optional, though people using ENS will want to create manifests. This is transparent in the CLI.

Added tests and client method in `client.go` to wrap the above.

### CODE REVIEW

**Review by nolash — CHANGES_REQUESTED** [review]
I'll keep comments general in this round, if you don't mind.

* I never liked mixing querystrings and formdata in POST requests, at least not in a way where the data that is being _saved_ is in querystrings. I really don't think anyone will be typing in this type of data by hand in a browser or anything. Can we make it formdata please?

* Can you please move the `GetDigest` method to the `mru.Request` object? That would make it easier to generate the digest from external code.

* Tracing the changes in the `binaryPut` and `binaryGet` functions from `GetDigest()` I had to go through 6 or was it 7 levels, several of which just add some data length check. I understand why you are building it in this way, but if the intention is clarity for the next developer, I'm not sure how much easier it's to get the overview when you have to leaf through so many files for what essentially is a serialization of lower to medium complexity.

* Also, I am sure you are aware that the cli is broken with the current code. That's ok for now since we're taking this in steps, but maybe we should merge to `mru-nometachunk` instead, to keep the deltas small and not merge in something that breaks functionality?

**Comment by nolash** [comment]
> Tracing the changes in the binaryPut and binaryGet functions from GetDigest() I had to go through 6 or was it 7 levels, several of which just add some data length check.

At a minimum, please give an overview on the top level of what the full serialization is, and what is being serialized where.

**Comment by jpeletier** [comment]
@nolash Thank you for your comments

> I never liked mixing querystrings and formdata in POST requests, at least not in a way where the data that is being saved is in querystrings. I really don't think anyone will be typing in this type of data by hand in a browser or anything. Can we make it formdata please?

This was a toughtful decision. This is not data that is posted to be stored as it usually happens with formdata. As of this PR, parameters in the query string are indeed query parameters. You are not specifying the content that will be stored, but where it is going to be stored: a URL in the purest sense: All those parameters are used to compute the lookup key where the data will be placed.

As a result, the same query string can be used to retrieve what you just posted. Just change `POST` with `GET` and you'll get the same stuff back. This is very intuitive.

Using formdata would mean either having to encode multipart, or encode the payload data in the formdata. Having the query parameters in the query string leaves the request body "clean" to stream up/down the binary payload.

> Can you please move the GetDigest method to the mru.Request object? That would make it easier to generate the digest from external code.

I will. I left a TODO in there for the whole `Request` / `SignedResourceUpdate`. I didn't want to do too much refactoring and drive you crazy as before.

> Tracing the changes in the binaryPut and binaryGet functions from GetDigest() I had to go through 6 or was it 7 levels, several of which just add some data length check. I understand why you are building it in this way, but if the intention is clarity for the next developer, I'm not sure how much easier it's to get the overview when you have to leaf through so many files for what essentially is a serialization of lower to medium complexity.

You're totally right. It is a bit of a pain now, but helps me during refactoring and the changes I am introducing as keeps concerns separated. Once we reach a more stable implementation, I will re-consolidate the serializers in a single place or think about this one again.

`doc.go` gives an overview, but it is a bit prosaic rather than providing a schematic.
Each structure has some comments on top that explain its purpose and layout.

> Also, I am sure you are aware that the cli is broken with the current code. That's ok for now since we're taking this in steps, [...] 

Not OK for me. I don't push untested / broken code intentionally and don't like to waste anyone's time. What is broken exactly in the CLI? It works for me.

> [...] but maybe we should merge to mru-nometachunk instead, to keep the deltas small and not merge in something that breaks functionality?

That is what I wanted to do, however I cannot open a PR against my own PR.
Can you create a new branch `mrudev` or something and give me write access to it (only to that branch, not the entire repo)? This would help greatly with this process and any upcoming changes.
Another option is take the debate to my fork under our GitHub account `epiclabs-io`, but that wouldn't be the best I think.
Worst case, you could create a branch, but if I don't have access to it I can't deal with rebases, etc, and depend on you to keep it updated.

File: swarm/api/http/server_test.go
Comment (jpeletier) [reply swarm/api/http/server_test.go]: fixed. thanks

File: swarm/api/http/server_test.go
Comment (jpeletier) [reply swarm/api/http/server_test.go]: fixed. thanks

**Comment by jpeletier** [comment]
Rebased after HTTP package refactor (PR 17309)

**Comment by nolash** [comment]
> Can you create a new branch mrudev or something and give me write access to it 

that doesn't seem to work :/ anyway, we shouldn't keep any long-running side branches if we can avoid it.

>  I left a TODO in there for the whole Request / SignedResourceUpdate. 

Good, let's do a separate PR just for that miniscule change. And a separate PR for the serialization refactor. In fact, let's try to keep future PRs as incremental and concrete as possible. This is an improvement from before :)

> What is broken exactly in the CLI? It works for me.

Hmm, maybe I was running the new branch against a node with old binary. Anyway, I pulled your recent rebase, and I got it to work now.

> As a result, the same query string can be used to retrieve what you just posted. Just change POST with GET and you'll get the same stuff back. This is very intuitive.

Ok, that make sense.

**Review by nolash — APPROVED** [review]

**Review by nolash — COMMENTED** [review]
We should PR this to `mru-nometachunk`. I take it we're waiting for an additional review on that one in order to merge, or..?

**Review by acud — CHANGES_REQUESTED** [review]
@jpeletier i commented on the modifications and additions to the `http` package in the code.

additionally, in your previous PR when I asked you to move the locator parameters to headers, you insisted that they should be in the locator because they refer to the location of the resource and that headers "aren't comfortable to work with" (or something along those lines). now you're moving them to the querystring but on what grounds? I fail to understand. The `GET`/`POST` swapping is not an acceptable argument in my world - MRUs will eventually have a `web3` wrapper and these implementation details would be abstracted into a javascript library thus transparent to the end user. If you really (still) insist on having this in the locator, I'd rather have it with named parameters chained after the scheme - then split that into string tuples and unmarshal into a struct.

Furthermore - the fact that you're offering both query-string params and a normal uri resource `meta` looks inconsistent.

File: swarm/api/http/server.go
Comment (jpeletier) [reply swarm/api/http/server.go]: Will do. Thanks.

File: swarm/api/http/server.go
Comment (jpeletier) [reply swarm/api/http/server.go]: Yes I know. I wasn't sure either where this method should go. I will move it to `api.go`.

As of this PR, a specific update of a specific MRU can be located with the values of `startTime`, `frequency`, `user`, `topic`, `period` `version`...

The manifest, which is needed for ENS, provides `startTime`, `frequency`, `user`, `topic`. So if you provide a manifest, you don't need to put these query parameters.

Providing manifest + period + version is what we already had

The reason I added the possibility of not having to provide the manifest is to save an extra lookup for applications that don't use manifests, e.g., the one we are building.

Moreover, with the new concept of MRU where you can deduct these parameters, manifests get in the way. For example, if I know your Ethereum address is 0xABC, I could lookup
`startTime=0&frequency=86400&user=0xABC&topic=myfeed` to see your Twitter-like feed of daily comments.
I could also lookup `startTime=0&frequency=86400&user=0xABC&topic=profilepic` and get your last profile picture, for example.
If you want to update your profile picture, just post the new one to `startTime=0&frequency=86400&user=0xABC&topic=profilepic`

or, if you had a manifest pointing to this in `profilepic.justelad.eth`, just POST the new one to `profilepic.justelad.eth` and you don't have to rememeber those parameters.

So the manifests are just a handy way of not having to remember `startTime`, `frequency`, `user`, `topic`. And you get a hash to a JSON understood by `bzz`. This makes it backward compatible with what we already had and all the ENS stuff.

I think both approaches are very useful.

**Comment by jpeletier** [comment]
> @justelad: additionally, in your previous PR when I asked you to move the locator parameters to headers, you insisted that they should be in the locator because they refer to the location of the resource and that headers "aren't comfortable to work with" (or something along those lines). now you're moving them to the querystring but on what grounds? I fail to understand. 

Interesting that you say *I asked you to move the locator parameters to headers*. So they are locators! 😉 

Yes, I said headers are not confortable to work with, but that was not my main point. That was supposed to be just a minor "nuisance point".
My final post on that matter is here: https://github.com/ethersphere/go-ethereum/pull/784#discussion_r203894517
I was never against query strings (`?name=value&name2=value2`). When given the choice of query strings vs REST-like path components, I did like path components more at that point, but query strings was nearer the consensus. I referred to this as a "cosmetic" taste. Unimportant.

As a summary of the above link, my first point was a semantic one:

- Headers: How. How you'd like your "what" delievered.
- URL: Specify what you want
- Body: the actual "what" that is being moved 

My second point was that putting stuff in custom headers is asking for trouble in my experience. In particular intermediate proxies filtering them out, browser CORS, preflights, and browser / intermediate proxy cache keys.

> The GET/POST swapping is not an acceptable argument in my world

This is not my argument but what the GET/POST definition is about... ¯\_(ツ)_/¯. According to w3c:
- *The GET method means retrieve whatever information (in the form of an entity) is identified by the Request-URI*
- *The POST method is used to request that the origin server accept the entity enclosed in the request as a new subordinate of the resource identified by the Request-URI*

Both talk about "resource identified by the Request-URI". As I explained earlier, our MRU is really identified by the URI with all those query string parameters.

> MRUs will eventually have a web3 wrapper and these implementation details would be abstracted into a javascript library thus transparent to the end user.

We will be writing that web3 wrapper (unless somebody does it first or wants to do it more). Precisely because of CORS and preflight requests we don't want to have to deal with stuff in custom headers in browsers.

> If you really (still) insist on having this in the locator, I'd rather have it with named parameters chained after the scheme - then split that into string tuples and unmarshal into a struct.

I insist. I am convinced this is the right way to go. But what do you mean with named parameters? isn't `?name=value&name2=value2` named parameters? This is what we already have in this PR.

> Furthermore - the fact that you're offering both query-string params and a normal uri resource meta looks inconsistent.

Query string vs. path components are not mutually exclusive. It is not one or the other.
For example, in GitHub:
https://github.com/ethersphere/go-ethereum/search?q=MRU
you can also go to 
https://github.com/ethersphere/go-ethereum
and that resource does not accept query parameters.

**Comment by jpeletier** [comment]
Te xgghf

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by acud** [comment]
> So they are locators!

I meant the path variables; surely the URLs are locators.

> query strings was nearer the consensus.

Who's and which consensus do you mean? Query string params are typically used in use cases when you would like to access _several_ items, or whenever you would like to do filtering and sorting (this is not my opinion, but a best practice). This is not the case here:
```
// bzz-resource://<id>/?period=n - get latest update on period n
// bzz-resource://<id>/?period=n&version=m - get update version m of period n
```
Both queries ask for a specific resource, thus I'd ask you to stick to your original implementation. Q.E.D.

> Both talk about "resource identified by the Request-URI". As I explained earlier, our MRU is really identified by the URI with all those query string parameters.

I understand. What I was referring to was that it seems like your motivation to implement this in a certain manner relates more to your terminal/browser habits rather than an API definition:
> As a result, the same query string can be used to retrieve what you just posted. Just change POST with GET and you'll get the same stuff back. This is very intuitive.

> But what do you mean with named parameters?

Path parameters: `POST bzz-resource:/period/1/version/5/...`
I'd rather see these as explicit params when looking for a specific resource, and, as mentioned before, query params for filtering/sorting.

File: swarm/api/http/server.go
Comment (acud) [reply swarm/api/http/server.go]: > The manifest, which is needed for ENS, provides startTime, frequency, user, topic. So if you provide a manifest, you don't need to put these query parameters.

It sounds to me like something that the client can do - so why put this logic into the proxy? you could just provide a simple formula on how to calculate this in javascript. I'd be more happy with a single-responsible endpoint and not an _all-arounder_ one.

Comment (jpeletier) [reply swarm/api/http/server.go]: > It sounds to me like something that the client can do - so why put this logic into the proxy? you could just provide a simple formula on how to calculate this in javascript. I'd be more happy with a single-responsible endpoint and not an all-arounder one.

I am not sure I understand. What do you mean that the client can calculate?

**Comment by jpeletier** [comment]
image

I don't like my comments deleted. What was it that caused my comment to be deleted? Thanks.

**Comment by acud** [comment]
> I don't like my comments deleted. What was it that caused my comment to be deleted? Thanks.

Check your phone's sent emails

**Comment by jpeletier** [comment]
oops, sorry about that. pocket text.

**Comment by jpeletier** [comment]
> Who's and which consensus do you mean?

See comments following https://github.com/ethersphere/go-ethereum/pull/784#discussion_r203894517 . @zelig and @nolash would prefer query string parameters, looks like.

> Query string params are typically used in use cases when you would like to access several items, or whenever you would like to do filtering and sorting

Not just for filtering and sorting, also to pinpoint a specific item:
GET /api/customers?id=5

The REST way would be:
GET /api/customers/5

Both are semantically correct. Whether to use one or the other is a matter of design / taste / pattern.

The thing is that our lookup key has more components than just one number. This makes it a bit longer. It is also very interesting that those components are clearly laid out.

> ```
> // bzz-resource://<id>/?period=n - get latest update on period n
> // bzz-resource://<id>/?period=n&version=m - get update version m of period n
> ```
> Both queries ask for a specific resource, thus I'd ask you to stick to your original implementation. Q.E.D.

Just to clarify, it was not my original implementation. It was already there.

The problem is that to pinpoint a specific resource we now have more components. If I followed that, this would look like this:

```
GET /bzz-resource:/topic/mytopic/owner/0xABCDABCDABCDABCDABCD/starttime/1231231/frequency/300/period/5/version/700
```
And this does not make sense, because while versions have periods and thus you could say that a version is a child of period, there is not a parent/child relationship among the rest of the parameters, so you can't put them in a hierarchical path layout. These are just the components of a search key.

**Comment by jpeletier** [comment]
> I understand. What I was referring to was that it seems like your motivation to implement this in a certain manner relates more to your terminal/browser habits rather than an API definition:

Please do not continue referring to my "personal terminal/browser habits" just because I made a side comment in a different PR. I am not contributing to design an API based on my habits. I already explained my motivations with strong evidence other than my personal preferences.

On a more philosophical point, I think just Web2 and Web3 are going to naturally collide precisely here in the frontier between them. It is just different worlds. There won't be a perfect solution.

In the meantime, as we iterate over the experimental concept of MRUs, it is easier to keep this as query string parameters because how search keys are built.

**Comment by acud** [comment]
> And this does not make sense, because while versions have periods and thus you could say that a version is a child of period, there is not a parent/child relationship among the rest of the parameters, so you can't put them in a hierarchical path layout. These are just the components of a search key.

RESTful path parameters do not always infer parent/child relationship, especially when describing graph-like data structures such as mutable resources, thus I don't see the path parameters even as a structure which has to conform to a certain order in this case. And yes, the discussion here is about taste/design style, rather than semantic correctness.

And just to put things in the open - I'm generally pushing towards REST when it comes to HTTP API design. I think it got enough traction in the wild and has refined itself as an industry standard and not without reason.

In any case I think we're in a deadlock here. If no-one else pitches into this discussion - leave it as it is.

**Review by acud — APPROVED** [review]

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 6d1e292eefa70b5cb76cd03ff61fc6c4550d7c36
  * Commits must be in the following format: **%{scope}: %{description}**

Guidelines are available at https://github.com/ethersphere/go-ethereum

---
This message was auto-generated by https://gitcop.com

**Comment by zelig** [comment]
I find @jpeletier points convincing. I side with query params and get/put symmetry.
@jpeletier would it make sense to support POSTing body  and put a hash only in the MRU data (`?use-reference=true` or something) 
Maybe later...

**Comment by jpeletier** [comment]
> I find @jpeletier points convincing. I side with query params and get/put symmetry.

I had a conversation later with @justelad on Gitter and we tabled this discussion since we both thought it was going nowhere. We came to the conclusion that this is still changing too frequently to spend time defining the perfect API now. For example, PR #881 does away with some of the parameters and adds new ones (time-based lookups) with different semantics. So the conclusion was let's move on and leave it as is so we can put the technology to use as soon as possible and get real user feedback.

From the discussion there a number of things I learned:
* Do not try to have the API do too many things. 
* See if we can push some of the actions to the client code.

For example, it makes sense to move manifest creation out and simplify the API as @justelad pointed out. This could be moved to `client.go` without changing the interface there.

> @jpeletier would it make sense to support POSTing body and put a hash only in the MRU data (?use-reference=true or something)

I think this would be one of those things we shouldn't do. :-). Also, to implement this, the client would still need to calculate the swarm hash of the content beforehand so it can sign the MRU with that hash inside. I would implement this functionality in `client.go` as a convenience method.

**Pull request closed without merging**
